weave/trace_server/clickhouse_trace_server_batched.py

```diff
@@ -456,6 +457,105 @@ def _expand_call_refs(
                     val["_ref"] = ref.uri()
                 set_nested_key(calls[i], col, val)
 
+    def _get_call_descendent_ids(
+        self,
+        project_id: str,
+        call_ids: list[str],
+        limit: Optional[int] = None,
+        depth: Optional[int] = None,
+        include_root: bool = False,
+    ) -> list[str]:
+        """
+        Helper function to get all descendant call ids of the requested root calls.
+
+        Args:
+            project_id: The project ID to search within
+            call_ids: List of root call IDs to find descendants for
+            limit: Optional maximum number of descendants to return
+            depth: Optional maximum depth of descendants to return (1 = immediate children only, 2 = children and grandchildren, etc.)
+            include_root: Whether to include the root calls in the result
+        """
+        if depth is not None and depth < 1:
+            raise ValueError("Depth must be a positive integer")
+
+        # Use recursive cte to get all children of requested calls, one batch
+        # at a time. First: parents, then children, then grandchildren, etc.
+        depth_condition = "" if depth is None else "AND d.depth < {depth:Int32}"
+        query = f"""
+        WITH RECURSIVE descendants AS (
+            -- Base case: get the root calls
+            SELECT id, parent_id, 0 as depth
+            FROM call_parts
+            WHERE project_id = {{project_id:String}}
+                AND id IN {{call_ids:Array(String)}}
+                AND deleted_at IS NULL
+
+            UNION ALL
+
+            -- Recursive case: get children of descendants
+            SELECT c.id, c.parent_id, d.depth + 1
+            FROM call_parts c
+            INNER JOIN descendants d ON c.parent_id = d.id
+            WHERE c.project_id = {{project_id:String}}
+                AND c.deleted_at IS NULL
+                {depth_condition}
+        )
+        SELECT DISTINCT id FROM descendants
+        LIMIT {{limit:Int32}}
+        SETTINGS allow_experimental_analyzer=1
```

Comment:
what does this do?

Reply:
Are there implications for non-clickhouse cloud users?  I don't think we have any now, but will we in future?

Reply:
clickhouse cloud is generally _stricter_ than other deployments, for example our dev setup does not require this extra setting